File: lib/py/uapi/internal/binary/ClientBinaryDecode.py
from typing import cast, TYPE_CHECKING
from uapi.internal.binary.BinaryEncoding import BinaryEncoding

if TYPE_CHECKING:
    from uapi.ClientBinaryStrategy import ClientBinaryStrategy


def client_binary_decode(message: list[object], recent_binary_encoders: dict[int, 'BinaryEncoding'],
                         binary_checksum_strategy: 'ClientBinaryStrategy') -> list[object]:
    from uapi.internal.binary.DecodeBody import decode_body
    from uapi.internal.binary.UnpackBody import unpack_body

    headers = cast(dict[str, object], message[0])
    encoded_message_body = cast(dict[str, object], message[1])
    binary_checksums: list[int] = cast(list[int], headers["bin_"])
    binary_checksum: int = binary_checksums[0]

    # If there is a binary encoding included on this message, cache it
    if "enc_" in headers:
        binary_encoding = cast(dict[str, int], headers["enc_"])
        new_binary_encoder = BinaryEncoding(binary_encoding, binary_checksum)

        recent_binary_encoders[binary_checksum] = new_binary_encoder

    binary_encoder: BinaryEncoding = recent_binary_encoders[binary_checksum]

    final_encoded_message_body = unpack_body(
        encoded_message_body) if headers.get("_pac") else encoded_message_body

    message_body = decode_body(final_encoded_message_body, binary_encoder)
    return [headers, message_body]

File: lib/ts/uapi/internal/binary/ClientBinaryDecode.ts
import { BinaryEncoding } from 'uapi/internal/binary/BinaryEncoding';
import { ClientBinaryStrategy } from 'uapi/ClientBinaryStrategy';
import { decodeBody } from 'uapi/internal/binary/DecodeBody';
import { unpackBody } from 'uapi/internal/binary/UnpackBody';

export function clientBinaryDecode(
    message: any[],
    recentBinaryEncoders: Record<number, BinaryEncoding>,
    binaryChecksumStrategy: ClientBinaryStrategy,
): any[] {
    const headers = message[0];
    const encodedMessageBody = message[1];
    const binaryChecksums: number[] = headers['bin_'] || [];
    const binaryChecksum = binaryChecksums[0];

    // If there is a binary encoding included on this message, cache it
    if ('enc_' in headers) {
        const binaryEncoding = headers['enc_'];
        const newBinaryEncoder = new BinaryEncoding(binaryEncoding, binaryChecksum);

        recentBinaryEncoders[binaryChecksum] = newBinaryEncoder;
    }

    binaryChecksumStrategy.updateChecksum(binaryChecksum);
    const newCurrentChecksumStrategy = binaryChecksumStrategy.getCurrentChecksums();

    for (const key in recentBinaryEncoders) {
        if (!(key in newCurrentChecksumStrategy)) {
            delete recentBinaryEncoders[key];
        }
    }

    const binaryEncoder = recentBinaryEncoders[binaryChecksum];

    let finalEncodedMessageBody: Record<any, any>;
    if (headers['_pac'] === true) {
        finalEncodedMessageBody = unpackBody(encodedMessageBody);
    } else {
        finalEncodedMessageBody = encodedMessageBody;
    }

    const messageBody = decodeBody(finalEncodedMessageBody, binaryEncoder);
    return [headers, messageBody];
}
